packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -238,106 +209,53 @@ contract Bridge is EssentialContract, IBridge {
-            }
+            _updateMessageStatus(msgHash, status);
         }
 
-        if (invocationDelay != 0 && msg.sender != proofReceipt[msgHash].preferredExecutor) {
-            // If msg.sender is not the one that proved the message, then there
-            // is an extra delay.
-            unchecked {
-                invocationDelay += invocationExtraDelay;
-            }
-        }
+        emit MessageExecuted(msgHash);
 
-        if (_isPostInvocationDelay(receivedAt, invocationDelay)) {
-            // If the gas limit is set to zero, only the owner can process the message.
-            if (_message.gasLimit == 0 && msg.sender != _message.destOwner) {
-                revert B_PERMISSION_DENIED();
-            }
+        refundAmount += _message.fee;
 
-            delete proofReceipt[msgHash];
-
-            uint256 refundAmount;
-
-            // Process message differently based on the target address
-            if (
-                _message.to == address(0) || _message.to == address(this)
-                    || _message.to == signalService
-            ) {
-                // Handle special addresses that don't require actual invocation but
-                // mark message as DONE
-                refundAmount = _message.value;
-                _updateMessageStatus(msgHash, Status.DONE);
-            } else {
-                uint256 gasLimit;
-                if (msg.sender == _message.destOwner) {
-                    // Use the remaining gas if called by a the destOwner, else
-                    // use the specified gas limit.
-                    gasLimit = gasleft();
-                } else {
-                    // The "1/64th rule" refers to the gasleft at the time the call is made. When a
-                    // contract makes a call to another contract, it can only forward 63/64 of the
-                    // gas remaining (gasleft) at that moment, ensuring that there is always some
-                    // gas reserved for the calling contract to complete its execution after the
-                    // called contract finishes. This does not necessarily relate to the gas amount
-                    // specified in the call itself, but rather to the actual remaining gas at the
-                    // time of the call.
-                    //
-                    // See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md
-                    if (_message.gasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
-
-                    gasLimit = _message.gasLimit;
-                }
-
-                if (_invokeMessageCall(_message, msgHash, gasLimit)) {
-                    _updateMessageStatus(msgHash, Status.DONE);
-                } else {
-                    _updateMessageStatus(msgHash, Status.RETRIABLE);
-                }
-            }
+        if (msg.sender != _message.destOwner) {
+            uint256 gasUsed = _calcGasToCharge(gas, GAS_OVERHEAD, _proof.length, numCacheOps);
```

Comment:
Hmm, why we need to calculate specifically with `_calcGasToCharge` with taking into account the number of caching ?

Isn't (psuedo) would to the job - so actually calculating the gas consumed:
```
uint256 gasWhenFunctionStarted = gasleft(); // same as like line 212
...
uint256 gasUsed = (gasWhenFunctionStarted - gasleft()) + gasOverhead
```

Reply:
Basically caching only benefit the relayer, not the message owner, therefore, if there is cache actions, the corresponding gas should be paid by the relayer, not the user.